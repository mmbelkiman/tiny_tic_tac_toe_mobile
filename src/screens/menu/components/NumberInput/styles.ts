import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  container: {flexDirection: 'row', alignItems: 'center'},
  textButton: {fontSize: 30},
  text: {
    fontSize: 32,
    textAlign: 'center',
  },
  textContainer: {
    minWidth: 110,
    height: 50,
    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: 8,
  },
});
