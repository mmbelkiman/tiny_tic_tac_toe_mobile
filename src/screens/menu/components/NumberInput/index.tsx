import React, {useState} from 'react';
import {Text, TouchableOpacity, View} from 'react-native';
import styles from './styles';

interface NumberInputProps {
  value: number;
  minValue: number;
  maxValue: number;
  onChangeValue: (newValue: number) => void;
  testID?: string;
}

const NumberInput: React.FC<NumberInputProps> = ({
  value,
  minValue,
  maxValue,
  onChangeValue,
  testID = '',
}) => {
  const [inputValue, setInputValue] = useState(value.toString());

  const handleMinusPress = () => {
    const newValue = Math.max(value - 1, minValue);
    setInputValue(newValue.toString());
    onChangeValue(newValue);
  };

  const handlePlusPress = () => {
    const newValue = Math.min(value + 1, maxValue);
    setInputValue(newValue.toString());
    onChangeValue(newValue);
  };

  return (
    <View style={styles.container}>
      <TouchableOpacity
        testID={`NumberInput${testID}.TouchableOpacity.minus`}
        onPress={handleMinusPress}>
        <Text style={styles.textButton}>[-]</Text>
      </TouchableOpacity>
      <View style={styles.textContainer}>
        <Text style={styles.text}>{`${inputValue}x${inputValue}`}</Text>
      </View>
      <TouchableOpacity
        testID={`NumberInput${testID}.TouchableOpacity.plus`}
        onPress={handlePlusPress}>
        <Text style={styles.textButton}>[+]</Text>
      </TouchableOpacity>
    </View>
  );
};

export default NumberInput;
