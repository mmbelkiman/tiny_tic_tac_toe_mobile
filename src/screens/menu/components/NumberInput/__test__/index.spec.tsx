import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import NumberInput from '../index';

describe('NumberInput component', () => {
  it('should render with the initial value', () => {
    const {getByText} = render(
      <NumberInput
        value={3}
        minValue={0}
        maxValue={5}
        onChangeValue={() => {}}
      />,
    );

    expect(getByText('3x3')).toBeTruthy();
  });

  it('should decrease the value when the minus button is pressed', () => {
    let currentValue = 3;
    const handleChangeValue = (value: number) => (currentValue = value);

    const {getByText} = render(
      <NumberInput
        value={currentValue}
        minValue={0}
        maxValue={5}
        onChangeValue={handleChangeValue}
      />,
    );

    const minusButton = getByText('[-]');
    fireEvent.press(minusButton);
    expect(currentValue).toBe(2);
  });

  it('should not decrease the value below the minimum value', () => {
    let currentValue = 0;
    const handleChangeValue = (value: number) => (currentValue = value);

    const {getByText} = render(
      <NumberInput
        value={currentValue}
        minValue={0}
        maxValue={5}
        onChangeValue={handleChangeValue}
      />,
    );

    const minusButton = getByText('[-]');
    fireEvent.press(minusButton);
    expect(currentValue).toBe(0);
  });

  it('should increase the value when the plus button is pressed', () => {
    let currentValue = 3;
    const handleChangeValue = (value: number) => (currentValue = value);

    const {getByText} = render(
      <NumberInput
        value={currentValue}
        minValue={0}
        maxValue={5}
        onChangeValue={handleChangeValue}
      />,
    );

    const plusButton = getByText('[+]');
    fireEvent.press(plusButton);
    expect(currentValue).toBe(4);
  });

  it('should not increase the value above the maximum value', () => {
    let currentValue = 5;
    const handleChangeValue = (value: number) => (currentValue = value);

    const {getByText} = render(
      <NumberInput
        value={currentValue}
        minValue={0}
        maxValue={5}
        onChangeValue={handleChangeValue}
      />,
    );

    const plusButton = getByText('[+]');
    fireEvent.press(plusButton);
    expect(currentValue).toBe(5);
  });

  it('renders correctly with initial value', () => {
    const {toJSON} = render(
      <NumberInput
        value={3}
        minValue={0}
        maxValue={5}
        onChangeValue={() => {}}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when decreasing value', () => {
    const {getByTestId, toJSON} = render(
      <NumberInput
        value={3}
        minValue={0}
        maxValue={5}
        onChangeValue={() => {}}
      />,
    );

    const minusButton = getByTestId('NumberInput.TouchableOpacity.minus');
    fireEvent.press(minusButton);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when increasing value', () => {
    const {getByTestId, toJSON} = render(
      <NumberInput
        value={3}
        minValue={0}
        maxValue={5}
        onChangeValue={() => {}}
      />,
    );

    const plusButton = getByTestId('NumberInput.TouchableOpacity.plus');
    fireEvent.press(plusButton);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when value reaches minimum', () => {
    const {getByTestId, toJSON} = render(
      <NumberInput
        value={0}
        minValue={0}
        maxValue={5}
        onChangeValue={() => {}}
      />,
    );

    const minusButton = getByTestId('NumberInput.TouchableOpacity.minus');
    fireEvent.press(minusButton);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when value reaches maximum', () => {
    const {getByTestId, toJSON} = render(
      <NumberInput
        value={5}
        minValue={0}
        maxValue={5}
        onChangeValue={() => {}}
      />,
    );

    const plusButton = getByTestId('NumberInput.TouchableOpacity.plus');
    fireEvent.press(plusButton);

    expect(toJSON()).toMatchSnapshot();
  });
});
