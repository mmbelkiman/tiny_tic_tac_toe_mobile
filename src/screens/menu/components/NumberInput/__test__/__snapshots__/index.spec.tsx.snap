// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput component renders correctly when decreasing value 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.minus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [-]
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "height": 50,
        "justifyContent": "center",
        "minWidth": 110,
        "paddingTop": 8,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 32,
          "textAlign": "center",
        }
      }
    >
      2x2
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.plus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [+]
    </Text>
  </View>
</View>
`;

exports[`NumberInput component renders correctly when increasing value 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.minus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [-]
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "height": 50,
        "justifyContent": "center",
        "minWidth": 110,
        "paddingTop": 8,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 32,
          "textAlign": "center",
        }
      }
    >
      4x4
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.plus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [+]
    </Text>
  </View>
</View>
`;

exports[`NumberInput component renders correctly when value reaches maximum 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.minus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [-]
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "height": 50,
        "justifyContent": "center",
        "minWidth": 110,
        "paddingTop": 8,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 32,
          "textAlign": "center",
        }
      }
    >
      5x5
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.plus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [+]
    </Text>
  </View>
</View>
`;

exports[`NumberInput component renders correctly when value reaches minimum 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.minus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [-]
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "height": 50,
        "justifyContent": "center",
        "minWidth": 110,
        "paddingTop": 8,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 32,
          "textAlign": "center",
        }
      }
    >
      0x0
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.plus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [+]
    </Text>
  </View>
</View>
`;

exports[`NumberInput component renders correctly with initial value 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.minus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [-]
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "height": 50,
        "justifyContent": "center",
        "minWidth": 110,
        "paddingTop": 8,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 32,
          "textAlign": "center",
        }
      }
    >
      3x3
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="NumberInput.TouchableOpacity.plus"
  >
    <Text
      style={
        {
          "fontSize": 30,
        }
      }
    >
      [+]
    </Text>
  </View>
</View>
`;
